[Core] Consolidate all node configurations into a new configs package - (Issue #396)

deblasis opened the pull request:
## Description

The main goal of this PR is to centralize configurations into a single `configs` package.
The feedback received was that the previous implementation was probably overly abstracted. Mockability is still preserved thanks to the fact that we are ultimately just dealing with structs without any logic attached to them.

It also tends to address:

```
// IMPROVE(#361): Design a better way to generate deterministic keys for testing.
```
https://github.com/pokt-network/pocket/compare/issue/396-consolidate-config?expand=1#diff-ee06bf613911836fdf48424397b7ec556c2e1a8ddba5adecca4c1782aea24087L30


## Issue

- Fixes #396
- Fixes #334
- Fixes #414


## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Centralized configs into the `configs` package
- Introduced a keygenerator that can work deterministically or randomly at will
- Centralize genesis as well, from two structs with overlapping states into a single one
- Simplified shared `core` types (Actor, Account, etc) under `coreTypes`
- Added missing `max_mempool_count` in config (it was causing P2P instabilities in LocalNet)

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`
<!--
If you added additional tests or infrastructure, describe it here.

Bonus points for images and videos or gifs.
-->

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

deblasis commented:
@Olshansk I would appreciate a quick scan on this, just to make sure it's getting where it's supposed to be. 

Please point out blatant gaps (apart from the fact that I need to centralize genesis as well but I am waiting for this initial feedback on the design before proceeding)

🙏

Olshansk reviewed (commented):
tl;dr I ❤️  where this is going

Olshansk commented on persistence/genesis.go:
```diff
@@ -44,97 +45,72 @@ func (m *persistenceModule) populateGenesisState(state modules.PersistenceGenesi
+	stakedActorsInsertConfigs := []struct {
+		Name     string
+		Getter   func() []*coreTypes.Actor
+		InsertFn func(address []byte, publicKey []byte, output []byte, paused bool, status int32, serviceURL string, stakedTokens string, chains []string, pausedHeight int64, unstakingHeight int64) error
```
```suggestion
		InsertFn func(address, publicKey, output []byte, paused bool, status int32, serviceURL, stakedTokens string, chains []string, pausedHeight, unstakingHeight int64) error
```

deblasis replied:
I would say yes but if we want to enforce a style I would say let's be consistent and scout the whole codebase for similar instances and rectify accordingly otherwise it becomes a mix and match.

Another thing we could do is adding newlines when the signatures are particularly long. I think you did that somewhere in the codebase and I like it since it allows the reader to avoid horizontal scrolling.

Let me know how you want to go about your suggestion and, more generally, about styling guidelines, perhaps this could be related to linting? (#295)

Olshansk replied:
```
I would say yes but if we want to enforce a style I would say let's be consistent and scout the whole codebase for similar instances and rectify accordingly otherwise it becomes a mix and match.
```

Agreed. I was trying to be consistent with line 83 of this file (just local consistency).

I believe we can enforce not specifying types (when they're the same) using the automatic linters @okdas set-up since these are standard golang suggestions. They're currently off because we need to do a one-time cleanup.

You are correct that #295 encompasses this.

```
Another thing we could do is adding newlines when the signatures are particularly long. I think you did that somewhere in the codebase and I like it since it allows the reader to avoid horizontal scrolling.
```

I do like it but think it's overkill here because the function is anonymous.

```
Let me know how you want to go about your suggestion and, more generally, about styling guidelines, perhaps this could be related to linting? (https://github.com/pokt-network/pocket/issues/295)
```

Let's just be consistent between this and line 83 and #295 will encompass the work of enforcing repo-wide guidelines.

Olshansk commented on persistence/genesis.go:
```diff
@@ -44,97 +45,72 @@ func (m *persistenceModule) populateGenesisState(state modules.PersistenceGenesi
-			log.Fatalf("an error occurred inserting staked tokens into %s pool: %s", types.PoolNames_FishermanStakePool.String(), err.Error())
-		}
+
+	stakedActorsInsertConfigs := []struct {
```
💯

Olshansk commented on persistence/genesis.go:
```diff
@@ -44,97 +45,72 @@ func (m *persistenceModule) populateGenesisState(state modules.PersistenceGenesi
+			Pool:     coreTypes.PoolNames_POOL_NAMES_APP_STAKE_POOL,
+		},
+		{
+			Name:     "node",
```
```suggestion
			Name:     "serviceNode",
```

TOL(Thinking Out Loud): I'm beginning to think we should rename these actors to `Servicers`

deblasis replied:
I agree, serviceNode is an improvement here since node could mean anything. If we want to rename them to Servicers I guess the only thing I would do is make sure that we are consistent in the codebase, documentation and perhaps even in Notion. Thoughts?

Olshansk commented on runtime/test_artifacts/keygenerator/keygen.go:
```diff
@@ -0,0 +1,66 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+var keygenInitOnce sync.Once
```
Optional NIT: we can use factoring here.

Aside: Big fan of this singleton pattern.

Olshansk commented on runtime/genesis.go:
```diff
@@ -4,43 +4,17 @@ import (
-}
-
-func parseGenesisJSON(genesisPath string) (g *runtimeGenesis, err error) {
+func parseGenesisJSON(genesisPath string) (g *genesis.GenesisState, err error) {
```
NIT: Can you rename this to `loadGenesis` and add a comment that `genesisPath` expects a JSON. 

`parseGenesisJSON(path)` feels incorrect.

Olshansk commented on persistence/test/setup_test.go:
```diff
@@ -80,17 +83,19 @@ func NewTestPostgresContext(t testing.TB, height int64) *persistence.PostgresCon
-		TxIndexerPath:  "",
-		TreesStoreDir:  "",
-	}))
+	teardownDeterministicKeygen := keygenerator.GetInstance().SetSeed(42)
```
This interface looks very clean IMO, so do we even need #414? Did you have something else in mind?

deblasis replied:
I have created #414 only to track the individual todo-item in #361 (I have used the feature "convert to issue") so that when this PR is merged, the todo item is automatically marked as completed and there are also breadcrumbs to here basically.

Olshansk commented on shared/core/types/proto/core_actor.proto:
```diff
@@ -0,0 +1,25 @@
+enum ActorType {
+  ACTOR_TYPE_UNSPECIFIED = 0;
+  ACTOR_TYPE_APP = 1;
+  ACTOR_TYPE_NODE = 2;
```
```suggestion
  ACTOR_TYPE_NODE = 2; // TODO: Consider rename `node` or `serviceNode` to `servicers` throughout the codebase.
```

Olshansk commented on runtime/test_artifacts/keygenerator/keygen.go:
```diff
@@ -0,0 +1,66 @@
+	return
+}
+
+func getPrivateKeySeed() (privateKeySeed int) {
```
Why did you expose this as a separate function rather than embedding the logic inside of `reset`? Could simplify the logic a bit.

```go
func GetInstance() *keyGenerator {
	if keygen == nil {
		keygenInitOnce.Do(func() {
			keygen = &keyGenerator{}
			keygen.reset()
		})
	}
	return keygen
}

// ...

func (k *keyGenerator) reset() {
	rand.Seed(timestamppb.Now().Seconds)
	k.privateKeySeed = rand.Int()
}
```

deblasis replied:
Fair point

Olshansk commented on persistence/genesis.go:
```diff
@@ -44,97 +45,72 @@ func (m *persistenceModule) populateGenesisState(state modules.PersistenceGenesi
+			if err != nil {
+				log.Fatalf("an error occurred inserting an %s in the genesis state: %s", saic.Name, err.Error())
+			}
+			if err = addValueToPool(coreTypes.PoolNames_POOL_NAMES_APP_STAKE_POOL.String(), act.GetStakedAmount()); err != nil {
```
```suggestion
			if err = addValueToPool(saic.Pool.String(), act.GetStakedAmount()); err != nil {
```

Olshansk commented on persistence/genesis.go:
```diff
@@ -44,97 +45,72 @@ func (m *persistenceModule) populateGenesisState(state modules.PersistenceGenesi
+				log.Fatalf("an error occurred inserting an %s in the genesis state: %s", saic.Name, err.Error())
+			}
+			if err = addValueToPool(coreTypes.PoolNames_POOL_NAMES_APP_STAKE_POOL.String(), act.GetStakedAmount()); err != nil {
+				log.Fatalf("an error occurred inserting staked tokens into %s pool: %s", coreTypes.PoolNames_POOL_NAMES_APP_STAKE_POOL, err.Error())
```
```suggestion
				log.Fatalf("an error occurred inserting staked tokens into %s pool: %s", coreTypes.Pool, err.Error())
```

deblasis replied:
I am not sure I understand this change

Olshansk commented on runtime/manager_test.go:
```diff
@@ -0,0 +1,365 @@
+		options       []func(*Manager)
+	}
+
+	buildConfigBytes, err := os.ReadFile("../build/config/config1.json")
```
NIT: Please move this path (and the one below) into constants at the top of the file with `Default` prefix

Olshansk commented on runtime/test_artifacts/generator.go:
```diff
@@ -2,149 +2,112 @@ package test_artifacts
-	for _, name := range typesPers.PoolNames_name {
-		if name == typesPers.PoolNames_FeeCollector.String() {
-			pools = append(pools, &typesPers.Account{
+func NewPools() (pools []*coreTypes.Account) { // TODO (Team) in the real testing suite, we need to populate the pool amounts dependent on the actors
```
```suggestion
 // REFACTOR: Test artifact generator should reflect the sum of the initial account values to populate the initial pool values
func NewPools() (pools []*coreTypes.Account) {
```

Olshansk reviewed (changes requested):
This PR is a 🕎 miracle!

Left a few minor comments throughout, but should be good to go right after. This is really going to help us start streamlining our work but also make the codebase much more understandable for anyone joining the codebase. Ty @deblasis!

deblasis commented on shared/core/types/proto/core_account.proto:
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
```
Totally! You made me also realise that the refactoring was incomplete. 🙏

deblasis commented on persistence/module.go:
```diff
@@ -8,25 +8,24 @@ import (
-	config       modules.PersistenceConfig
-	genesisState modules.PersistenceGenesisState
+	config       *configs.PersistenceConfig
+	genesisState *genesis.GenesisState
```
Fair point. However, I guess we need to think about pros and cons. 
At a glance a con could be the fact that in tests we would have to mock the bus every time. (of course it's a one-off but still). 
So basically we are enforcing the bus as dependency. Makes sense if we think about how the whole thing works but what are the benefits in not having pointers to the config and genesis inside the modules?

Olshansk replied:
Ditto - fair point.

Given what you said, I don't love either approach but prefer what we have for now over the alternative.

deblasis commented:
@Olshansk:
> This PR is a 🕎 miracle!
> 
> Left a few minor comments throughout, but should be good to go right after. This is really going to help us start streamlining our work but also make the codebase much more understandable for anyone joining the codebase. Ty @deblasis!

LOL, too kind! I have addressed the majority of the comments, left some open questions (that might lead to creating separate issues) and I have to update the CHANGELOGs but I guess it's ready for another pass :)

Olshansk commented on utility/types/proto/stake_status.proto:
```diff
@@ -0,0 +1,11 @@
+option go_package = "github.com/pokt-network/pocket/utility/types";
+
+// DISCUSS: *Design Decision* deprecating StakeStatus
+enum StakeStatus {
```
```suggestion
REFACTOR(#258): Rename the protobuf types to follow best practices
enum StakeStatus {
```

Olshansk reviewed (changes requested):
It's a pretty big PR so hard to catch small errors. Can you run the [iteration 3 demo](https://github.com/pokt-network/pocket/blob/main/docs/demos/iteration_3_end_to_end_tx.md) to see if everything checks out and still works?

Also, at some point (not in this PR) we should update the main READMEs around our code structure, but I think we have some more cleanup to do first.

deblasis commented:
@Olshansk 
> It's a pretty big PR so hard to catch small errors. Can you run the [iteration 3 demo](https://github.com/pokt-network/pocket/blob/main/docs/demos/iteration_3_end_to_end_tx.md) to see if everything checks out and still works?

100%

> Also, at some point (not in this PR) we should update the main READMEs around our code structure, but I think we have some more cleanup to do first.

Agreed

deblasis commented:
> @Olshansk
> 
> > It's a pretty big PR so hard to catch small errors. Can you run the [iteration 3 demo](https://github.com/pokt-network/pocket/blob/main/docs/demos/iteration_3_end_to_end_tx.md) to see if everything checks out and still works?
> 
> 100%

I ran the iteration demo and also updated it 

https://github.com/pokt-network/pocket/pull/403/files#diff-79ce7fc016a2250751d0d0f7a5621957e9feb4f4971e497ff49ad50d3f547899R100-R144

It looks like it working. The only weird thing / **stupid question** that I have noticed is that if I send 1000 POKT I see that the DAO takes 9000 POKT commission?

```bash
go run app/client/*.go --path_to_private_key_file=./val1.json Account Send 6f66574e1f50f0ef72dff748c3f11b9e0e89d32a 67eb3f0a50ae459fecf666be0e93176e92441317 1000
```
![image](https://user-images.githubusercontent.com/29378614/209881365-60e76014-6400-4f22-94cd-a75897fbf698.png)

![image](https://user-images.githubusercontent.com/29378614/209881384-4de69a13-177d-45f0-9555-a495aaa4e754.png)

If the answer is simply: inflation we are all good, just wanted to make sure I get it. 🙂


> 
> > Also, at some point (not in this PR) we should update the main READMEs around our code structure, but I think we have some more cleanup to do first.
> 
> Agreed

I have updated the CHANGELOG. If you don't flag other major changes I'll update the readmes

Olshansk commented on runtime/configs/config.go:
```diff
@@ -12,3 +14,64 @@ type Config struct {
+	return cfg
+}
+
+func WithPK(pk string) func(*Config) {
```
```suggestion
// There is an option to configure module-specific keys in order to enable different "identities"
// for different purposes (i.e. validation, P2P, servicing, etc...).
func WithPK(pk string) func(*Config) {
```

Olshansk commented on docs/demos/iteration_3_end_to_end_tx.md:
```diff
@@ -96,12 +97,32 @@ Show all the commands available in the CLI:
 
+#### Accounts setup
+
+Since our Keybase is under development, currently we have to manually inject the private keys of the accounts we want to use in the CLI.
```
```suggestion
Since our Keybase is under development, we have to manually inject the private keys of the accounts we want to use in the CLI.
```

Olshansk commented on docs/demos/iteration_3_end_to_end_tx.md:
```diff
@@ -96,12 +97,32 @@ Show all the commands available in the CLI:
+
+Since our Keybase is under development, currently we have to manually inject the private keys of the accounts we want to use in the CLI.
+
+For the following steps you'll need to use the accounts of the first two validators, therefore you have some options:
```
```suggestion
For the following steps, you'll need to use the accounts of the first two validators in the hard-coded development genesis file. Therefore you have some options:
```

deblasis commented on runtime/manager_test.go:
```diff
@@ -91,275 +244,46 @@ func TestNewManagerFromReaders(t *testing.T) {
 				clock.New(),
 			},
+			assertion: func(tt require.TestingT, want, got interface{}, _ ...interface{}) {
+				require.Equal(tt, want.(*Manager).config.P2P, got.(*Manager).config.P2P)
```
I selected a configuration value I was interested in and used it as a "probe" to prove the concept and that the behaviour is as expected.

Do you think is necessary to add the same test for all the other modules as well?
Perhaps I should name the test case more generically like "an unset configuration value should fallback to its default"

deblasis commented on shared/core/types/pool_names.go:
```diff
@@ -0,0 +1,18 @@
+package types
```
Done, for consistency, I renamed the enum as well

deblasis commented on persistence/docs/CHANGELOG.md:
```diff
@@ -5,19 +5,91 @@ All notable changes to this module will be documented in this file.
-## [0.0.0.10] - 2022-12-21
+## [0.0.0.17] - 2022-12-28
+
+- Added missing `ActorType` in `GetAllXXXX()` functions
```
It was never there to begin with, it was a bug that I am attempting to fix.

`getActor` returns the actor with no notion of its type. 

https://github.com/pokt-network/pocket/blob/main/persistence/shared_sql.go#L85-L110

deblasis replied:
I think I can genericize it more not that we have `coreTypes` 😉, I can leverage the `actorSchema`.

Apart from that, I am confident that I am not breaking anything, the type was unspecified before 🙂

PTAL: https://github.com/pokt-network/pocket/pull/403/commits/d7a521eb816c33d810f1e29f848028727512a149

Olshansk replied:
Left a comment in `persistence/shared_sql.go`

deblasis commented on p2p/CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## [0.0.0.14] - 2022-12-28
+
+- `ActorsToAddrBook` now skips actors that are not validators since they don't have a serviceUrl generic parameter
```
Currently, we use the serviceUrl to connect to peers and to send messages to them, I believe that the `serviceUrl` is likely gonna be replaced by something like a LibP2P Address

https://docs.libp2p.io/concepts/fundamentals/addressing/

Eg: 
`/ip4/7.7.7.7/tcp/4242/p2p/QmRelay/p2p-circuit/p2p/QmRelayedPeer`

but that's going to be dynamic and transient and I don't see how it would be part of the state.

That said, I am not sure we should change the genesis state at all to handle addressing. 
It's the addressing that has to change instead.

Also, what is this concept of generic parameter? Where can I learn more about it?

Olshansk replied:
Regarding libp2p - makes sense when we get there.

Regarding the val filtering - see the comment I left in `p2p/addrbook_provider/persistence.go`

-----

`Also, what is this concept of generic parameter? Where can I learn more about it?`

This is just a poor design decision that was the result of a compromise during the shared module refactor.

We (currently) have 4 protocol actor types:
- Fisherman
- Validator
- Servicer
- Application

There are things they all share (e.g. an address) and there are things that are specific:
- Servicers need a "serviceURL" to service nodes
- Fisherman need a "serviceURL" to enable QoS
- Applications need a 'maxRelaysCount' which is only specific to them

The two potential future paths we take are:
1. One Protobuf per actor
2. One `BaseActor` protobuf that's embedded in each ProtocolActor type

See #313 for the origin document.

deblasis replied:
Thanks for the wholesome explanation and context🙏

Olshansk replied:
🙏

deblasis commented on shared/CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## [0.0.0.8] - 2022-12-28
+
+- Added `PoolNames.FriendlyName` method
```
Is it? Nice then! It sounded "friendly" enough and I honestly didn't know how else to name this.

Olshansk commented on shared/core/types/pools.go:
```diff
@@ -0,0 +1,18 @@
+	poolFriendlyNames = map[Pools]string{
+		Pools_POOLS_UNSPECIFIED:        "Unspecified",
+		Pools_POOLS_DAO:                "DAO",
+		Pools_POOLS_FEE_COLLECTOR:      "FeeCollector",
```
```suggestion
		Pools_POOLS_FEE_COLLECTOR:      "FeeCollectorPool",
```

deblasis replied:
Are you sure you want to change this? The current values reflect what's currently on `main`, also, I could argue that they are entities in the table `pool` so the suffix is redundant.

Olshansk commented on shared/core/types/pools.go:
```diff
@@ -0,0 +1,18 @@
+func init() {
+	poolFriendlyNames = map[Pools]string{
+		Pools_POOLS_UNSPECIFIED:        "Unspecified",
+		Pools_POOLS_DAO:                "DAO",
```
```suggestion
		Pools_POOLS_DAO:                "DAOPool",
```

deblasis replied:
https://github.com/pokt-network/pocket/pull/403#discussion_r1060922509

Olshansk commented on persistence/shared_sql.go:
```diff
@@ -92,7 +92,7 @@ func (p *PostgresContext) getActor(actorSchema types.ProtocolActorSchema, addres
 		return
 	}
-
+	actor.ActorType = actorSchema.GetActorType()
```
`getActor` uses `getActorFromRow` under the hood

BUT

`GetAllXXX` uses `getActorFromRow` directly

So this won't actually fully work.

I'm okay with leaving it as is because we have a ton of scope creep, but could you add a TODO to fix it (in a follow up PR). I also think this is a really good opportunity to do some TDD to be future-proof.

deblasis replied:
I coudn't resist... https://github.com/pokt-network/pocket/pull/403/commits/fb42810947fe706561077ad191ee6a81246dab43 and https://github.com/pokt-network/pocket/pull/403/commits/6ff3c3ad2b446443df150931103c1b2b60df562b

Olshansk commented on p2p/addrbook_provider/persistence.go:
```diff
@@ -69,9 +71,13 @@ func (pabp *persistenceAddrBookProvider) GetStakedAddrBookAtHeight(height uint64
 	book := make(typesP2P.AddrBook, 0)
 	for _, v := range actors {
+		// only add validator actors since they are the only ones having a service url in their generic param at the moment
+		if v.ActorType != coreTypes.ActorType_ACTOR_TYPE_VAL {
```
Related to one of my other comments, fisherman and serviceNodes also have serviceURLs in their parameter, and can (and should) also participate in structured broadcast.

Should they not be included as well?

![Screenshot 2023-01-03 at 11 49 05 AM](https://user-images.githubusercontent.com/1892194/210402625-0a00fcb1-2c66-42a9-b877-f2d1fff247ce.png)

deblasis replied:
TL;DR: The way I see it, **NO**... 

Like you [said](https://github.com/pokt-network/pocket/pull/403#discussion_r1060764619), it looks like these `serviceUrl`s have a different meaning depending on the actor type and therefore using them to connect to a node for consensus related messaging/structured broadcast is a... shortcut.

Nodes are listening on the `ConsensuPort` which by default is 8080 and it seems that `serviceUrl` has been used to have an easy way to address them.

I guess this is because we don't have -yet- a full-fledged P2P module that owns the addressbook and can be relied upon for routing and messaging (that will come with peer discovery and churn).

We won't address messages via `serviceUrl` but rather by `address` or `nodeId`, then P2P would figure out how to reach that peer.

In `LocalNet` we spin up 4 validator nodes and the debug commands are all related to consensus.
This explains why previously we had 👇👈 while now we have 👉👇
![image](https://user-images.githubusercontent.com/29378614/210437954-08ad4cba-acec-49a5-9801-89ba9765ae05.png)

If we consider other actors I can only imagine that since we have overlaps (eg: `node1.consensus:8080` in `genesis.json` is a Validator, a ServiceNode and also a Fisherman) we would have duplicate messages at a minimum.

I wish Brian was here to pick his brain on this :) the ways I see it, structured gossip happens under the hood in the P2P module, when you say "broadcast to all validators" you don't know/care if the message is also handled and retransmitted by a Fisherman in order to reach a "nearby" Validator. That's implementation detail of the P2P algorithm.

I would say that for now we have to filter. Happy to discuss further even offline. 

Just let me know how to proceed here.

Olshansk replied:
Solid explanation. I'm convinced.

Excited to see what our next steps are going to be here :)

deblasis replied:
Added https://github.com/pokt-network/pocket/issues/426 PTAL 🙏

Olshansk commented on val1.json:
```diff
@@ -0,0 +1 @@
+"6fd0bc54cc2dd205eaf226eebdb0451629b321f11d279013ce6fdd5a33059256b2eda2232ffb2750bf761141f70f75a03a025f65b2b2b417c7f8b3c9ca91e8e4"
```
Intentional?

deblasis replied:
nope. Genuine mistake

Olshansk commented on persistence/test/actor_test.go:
```diff
@@ -0,0 +1,38 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetAllStakedActors(t *testing.T) {
```
🙌

Olshansk reviewed (changes requested):
One tiny comment about the addition of `val1.json` and `val2.json` to the root dir.

1. Add it to `.gitignore`
2. Update `docs/demos/iteration_3_end_to_end_tx.md` to move it into `/tmp` and reference that directly.

Olshansk reviewed (approved):
LFG!

🚢 🚢🚢
